pytorch_lightning/plugins/training_type/ipu.py

```diff
@@ -0,0 +1,336 @@
+        """
+        super().__init__(parallel_devices, cluster_environment)
+        if not _POPTORCH_AVAILABLE or not poptorch.ipuHardwareIsAvailable():
+            raise MisconfigurationException(
+                "The IPU Accelerator requires IPU devices to run. "
+                "Learn more or get started with IPUs at https://www.graphcore.ai/getstarted"
+            )
+
+        self.convert_model_to_half = convert_model_to_half
+        self.device_iterations = device_iterations
+        self.autoround_num_ipus = autoround_num_ipus
+        self.autoreport = autoreport
+        self.autoreport_dir = autoreport_dir
+        self.poptorch_models = {}
+        self._original_accumulate_grad_batches = None
+
+        if self.autoreport:
+            options = {"autoReport.all": self.autoreport}
+            if self.autoreport_dir:
+                self._fs = get_filesystem(str(self.autoreport_dir))
+
+                if not self._fs.exists(self.autoreport_dir):
+                    self._fs.makedirs(self.autoreport_dir)
+                options["autoReport.directory"] = self.autoreport_dir
+            os.environ["POPLAR_ENGINE_OPTIONS"] = json.dumps(options)
+
+    @property
+    def lightning_module(self) -> Optional[LightningModule]:
+        return self.model.module if isinstance(self.model, LightningIPUModule) else self.model
+
+    def pre_dispatch(self) -> None:
+        self._handle_gradient_accumulation_steps()
+        if self.convert_model_to_half:
+            log.info('Using full 16bit precision, converting LightningModule weights to FP16.')
+            self.model = self.model.half()
+        precision = self.lightning_module.trainer.precision
+        precision = 16 if self.convert_model_to_half else precision
+
+        model = LightningIPUModule(self.lightning_module, precision)
+        self.model = model
+
+        # Separate models are instantiated for different stages, but they share the same weights on host.
+        # When validation/test models are run, weights are synced first.
+
+        if self.lightning_module.trainer.state.stage is RunningStage.TRAINING:
+            # Create model for training which will run training.
+            optimizer = self.lightning_module.trainer.optimizers[0]
+            model = poptorch.trainingModel(model=model, options=self._create_opts(training=True), optimizer=optimizer)
+            self.poptorch_models[RunningStage.TRAINING] = model
+        for x in (RunningStage.VALIDATING, RunningStage.TESTING, RunningStage.PREDICTING):
+            model = poptorch.inferenceModel(
+                model=model,
+                options=self._create_opts(training=False),
+            )
+            self.poptorch_models[x] = model
+
+    @property
+    def replication_factor(self):
+        return len(self.parallel_devices)
+
+    def _create_opts(self, training: bool):
+        opts = poptorch.Options()
```

Comment:
One issue is that there are some quite important options which some users might want to change in that struct. Could that be exposed to the user somehow? Maybe as part of the `IPUPlugin` class? It would mean that users who change the options have to pass that in manually (instead of just `ipu=`) but maybe that's ok?

Reply:
Great suggestion! We should allow the user to pass in an opts if they'd like! I think because of the differences, we might want to allow the user to pass in an opts for training and for inference, does that make sense?

Reply:
Yeah I think that makes sense